cli/file_fetcher.rs

```diff
@@ -393,17 +393,40 @@ impl SourceFileFetcher {
     let download_job = self.progress.add("Download", &module_url.to_string());
     let dir = self.clone();
     let module_url = module_url.clone();
+    let headers = self.get_source_code_headers(&module_url);
+    let module_etag = headers.etag;
 
     // Single pass fetch, either yields code or yields redirect.
     let f = async move {
-      match http_util::fetch_string_once(&module_url).await? {
+      match http_util::fetch_string_once(&module_url, module_etag).await? {
+        FetchOnceResult::NotModified => {
+          let maybe_source_file = dir.fetch_cached_remote_source(&module_url);
+
+          // Explicit drop to keep reference alive until future completes.
+          drop(download_job);
+
+          match maybe_source_file {
+            Ok(Some(source_file)) => Ok(source_file),
+            _ => {
+              dir
+                .fetch_remote_source_async(
+                  &module_url,
+                  use_disk_cache,
+                  cached_only,
+                  redirect_limit - 1,
+                )
+                .await
+            }
+          }
```

Comment:
I don't understand this `match` clause. If you pass down `module_etag` that means that we already have that module fetch and cached locally. So I think we should unwrap on `maybe_source_file`